src/Engine/ProtoCore/AssociativeGraph.cs

```diff
@@ -961,6 +961,36 @@ public void PushChildNode(GraphNode child)
             child.ParentNodes.Add(this);
         }
 
+        internal void ClearCycles(IEnumerable<GraphNode> graphNodes)
+        {
+            Stack<GraphNode> stack = new Stack<GraphNode>();
```

Comment:
this looks like DFS - is there any benefit to this traversal algorithm rather than BFS or it should not matter what order traversal happens in as long as we iterate over all downstream nodes?

Reply:
In general each graphnode has just one child node so I guess it shouldn't benefit much if we use BFS plus as you said we need to touch each child node once including the ones that are the deepest. In most cases the dependency graph is actually just a chain of dependencies so the most optimal way would have been a simple linear search but since I need to account for multiple dependents in the most general case I chose DFS.

Reply:
Could be less code if we make it recursive , but I remember you are not a fan :)